### DISCUSSION TITLE
Benchmarking and Profiling

### DISCUSSION THREAD

**Discussion by cheshmi** [OpenCilk/opencilk-project#171] [General]
I am interested in profiling the memory performance of OpenCilk programs. I know that Cilkscale can be used to analyze the load balance and parallelism of OpenCilk programs. Is there a similar tool that can measure the cache or TLB accesses of OpenCilk programs? And is the Cilkscale for any parallel shared-memory programming model?

Thanks,
Kazem

**Comment by neboat** [comment]
Hey Kazem,

We don't currently have OpenCilk-specific memory-analysis tools as part of an official OpenCilk release.

What kind of memory analysis are you interested in doing?

In practice, I've used several existing tools to analyze memory accesses in OpenCilk programs.
- I have often used `perf` on systems where performance counters are available for measuring cache accesses, TLB accesses, and other memory operations (e.g., prefetch operations).  I have found that `perf` works well even when OpenCilk programs execute in parallel (on multiple worker threads), but performance counters can be quite microarchitecture dependent and often produce nondeterministic results that can be difficult to understand.
- I have used `cachegrind` to annotate an OpenCilk program with the number of memory accesses performed.  However, `cachegrind`, like other tools in the `valgrind` suite, might require running the OpenCilk program on 1 worker thread.
- When I'm more interested in profiling memory *allocation*, I have used the heap profiler in Google `pprof`.  That heap profiler requires the use of `tcmalloc`, which I have generally found to be a quality fast parallel memory allocator, regardless of whether or not I'm profiling heap usage.
- I have used Google's AddressSanitizer to check OpenCilk programs for memory errors.  We added support to the OpenCilk runtime to let users run AddressSanitizer on their OpenCilk programs, even in parallel.  As such, I like AddressSanitizer for quickly checking OpenCilk programs for memory errors.

We have also done research work to analyze memory usage of Cilk programs, specifically, to analyze the memory high-water mark of a Cilk program on different processor counts.  That tool is called Cilkmem, but I don't think we have a production release of that tool right now.

I'm not sure I understand your second question, about whether Cilkscale is "for any parallel shared-memory programming model."  Can you elaborate?

Cheers,
TB

**Comment by cheshmi** [comment]
Hey TB,

I appreciate your clear explanation. 

- I agree that `perf` is the best option when counters are available, but I couldn't figure out how to use it with instrumentation to profile a specific part of the code. Did you manage to do that?

- I also used `cachegrind` a lot and found it helpful, but I eventually switched to real counters.

- I haven't tried `pprof` and addressSanitizer. I think `valgrind` covered my needs in this area.

I was wondering if other programming models, such as OpenMP, are compatible with Cilkscale. But I understand that Cilkscale is designed for OpenCilk only.

Cheers,
Kazem
